uvicorn/_async_agnostic/backends/base.py

```diff
@@ -0,0 +1,160 @@
+import signal
+import socket
+from typing import Any, AsyncContextManager, Awaitable, Callable, List, Optional, Tuple
+
+from ...config import Config
+from ..state import ServerState
+
+
+class Event:
+    async def set(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    def is_set(self) -> bool:
+        raise NotImplementedError  # pragma: no cover
+
+    async def wait(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    def clear(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+
+class Queue:
+    async def get(self) -> Any:
+        raise NotImplementedError  # pragma: no cover
+
+    async def put(self, item: Any) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    async def aclose(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+
+class AsyncSocket:
+    def get_local_addr(self) -> Optional[Tuple[str, int]]:
+        raise NotImplementedError  # pragma: no cover
+
+    def get_remote_addr(self) -> Optional[Tuple[str, int]]:
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def is_ssl(self) -> bool:
+        raise NotImplementedError  # pragma: no cover
+
+    async def read(self, n: int) -> bytes:
+        raise NotImplementedError  # pragma: no cover
+
+    async def write(self, data: bytes) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    async def send_eof(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    async def aclose(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def is_closed(self) -> bool:
+        raise NotImplementedError  # pragma: no cover
+
+
+class AsyncListener:
+    @property
+    def socket(self) -> socket.SocketType:
+        raise NotImplementedError  # pragma: no cover
+
+
+class TaskStatus:
+    IGNORED: "IgnoredTaskStatus"
+
+    def __init__(self, event: Event) -> None:
+        self._value_event = event
+
+    async def started(self, value: Any = None) -> None:
+        self._value = value
+        await self._value_event.set()
+
+    async def get_value(self) -> Any:
+        await self._value_event.wait()
+        assert hasattr(self, "_value")
+        return self._value
+
+
+class TaskHandle:
+    async def cancel(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+
+class IgnoredTaskStatus(TaskStatus):
+    def __init__(self) -> None:
+        super().__init__(None)  # type: ignore
+
+    async def started(self, value: Any = None) -> None:
+        pass
+
+    async def get_value(self) -> Any:
+        return None
+
+
+TaskStatus.IGNORED = IgnoredTaskStatus()
+
+
+class AsyncBackend:
```

Comment:
This "backends" directory follows the same approach than we did in HTTPCore. It doesn't use `anyio` directly — although that would be a valid option — so that I have full control over the APIs I need for building this out. I'm still wary of requiring `anyio` to be added as an extra dependency, rather than us just writing down some glue code suited to our exact needs.